PelicanPlatform/pelican
# Add Oauth2 client with CILogon

Close #355 

* Added a generic OAuth2 client to do third-party login for servers, with default provider being CILogon
* Enabled OAuth2 for `registry` server

## Testing Instructions

**PLEASE NOTE THE TEST INSTRUCTIONS ARE UPDATED AS OF 2023/12/08**

* Make sure you have `OIDC.ClientID` configured to your CILogon client ID
* Also ensure that you created a file under `/etc/pelican/oidc-client-secret` and copy-paste your  CILogon client secret there
* Also ensure that you either ask Cannon to register your redirect URL at CILogon as  `https://<registry-host>/api/v1.0/auth/cilogon/callback` or set your namespace registry web port to be `9999`
* Spin up `registry`: `./pelican registry serve -p 9999`
* Go to `https://<registry-host>/api/v1.0/auth/cilogon/login` and you should be redirected to CILogon for authentication
* After you authenticated, the page should be redirected back to `https://<registry-host>/view/`, which should be a server selection page and if you pick any server (say origin or director), you should be able to see their dashboard page without need to "login" 
* This auth flow supports `next_url` which can take you to where-ever you were before redirected to login. To test this function, go to `https://<registry-host>/api/v1.0/auth/cilogon/login?next_url=%2Fview%2Fconfig` and you should be eventually taken back to `/view/config/`
